core/templates/dev/head/pages/exploration-player-page/suggestion-modal-for-learner-local-view/suggestion-modal-for-exploration-player.service.ts

```diff
@@ -50,6 +50,25 @@ angular.module('oppia').factory('SuggestionModalForExplorationPlayerService', [
             var stateName = PlayerPositionService.getCurrentStateName();
             var displayedCard = PlayerTranscriptService.getCard(
               PlayerPositionService.getDisplayedCardIndex());
+            var componentRe = (
+              /(<(oppia-noninteractive-(.+?))\b[^>]*>)[\s\S]*?<\/\2>/g
+            );
+            var wrapComponents = function(html) {
+              if (html === undefined) {
```

Comment:
This all seems like it's duplicating code from ck-editor-4-rte.directive.ts. We should never do that: cf. the DRY principle.

I think you need to find a way to somehow fetch the relevant data from the CKEditor directive or implement something that gives you the relevant data, without duplicating all this code here. I suggest chatting with @bansalnitish about this since he implemented this originally and should be able to help you.

Reply:
@seanlip , I'll try a different approach then after talking to @bansalnitish .

Reply:
I would suggest add the common code in a new function in `exploration-html-formatter.service.ts` (I am fine with having a new service too) and then use the function both here and in `ck-editor-4-rte.directive.ts`.

Reply:
@bansalnitish @seanlip,I was thinking of another approach,we can compare `originalHtml` and `suggestionHtml` by undoing the effect added by the function `wrapComponents`,i.e remove the extra div or span tags added by the `wrapComponent` function and then compare. would this be a proper way of fixing this.

Reply:
I wonder if there's any way you can get the original data (or what CKEditor thinks is the original data, without transformations)?

I ask because, at some point, we take the stuff from CKEditor and save it to the backend. So somewhere there should be a place where CKEditor exposes/exports the data to be readable "from the outside". If we take that html string and compare it with the other one, that would probably make the most sense conceptually. (I.e. we should be comparing the "raw" html strings that would be stored in the datastore, not html strings with CKEditor transformations/formatting added to them.)

Reply:
@seanlip ,Could you specify where exactly in the codebase the data in CKEditor saved to the backend. I couldn't find anything in `ck-editor-4-rte.directive.ts. Thanks

Reply:
I'm afraid I don't know -- you'll need to debug. But, to clarify what I said, I'm not saying that the CKEditor saves the data to the backend. I'm saying that, at some point between CKEditor and the backend, the data gets converted to its "canonical form", and you should try to get hold of the data at that point, not while it's still in ckeditor-land. You will probably need to dig through the flows to see where that data is accessible, though.

If you get stuck, I think @bansalnitish can help; he knows more about the RTE than I do at this point!

Reply:
@bansalnitish ,@seanlip , Is this the canonical form of data we were talking about?
This is the cleaned data Which gets loaded into CKEditor at line 235 inside the `onChange` function(ck-editor-4-rte.directive.ts)
![canicnal](https://user-images.githubusercontent.com/31371571/69792988-a1eff380-11ed-11ea-9236-70a339202381.png)
After inspecting this data, I think it has all the data needed without any CKEditor modifications.
Should we use this data for comparing with originalHtml?

Reply:
Looks reasonable on the surface, but if you want to be absolutely sure, I suggest that you check with the examples in the data/explorations folder in the codebase.